subsys/bluetooth/host/adv.c

```diff
@@ -392,7 +391,131 @@ static int hci_set_ad_ext(struct bt_le_ext_adv *adv, uint16_t hci_op,
 	return bt_hci_cmd_send_sync(hci_op, buf, NULL);
 }
 
-static int set_ad(struct bt_le_ext_adv *adv, const struct bt_ad *ad,
+static int hci_set_adv_ext_fragment(const struct bt_le_ext_adv *adv, uint16_t hci_op,
+				    const struct bt_data *data, bool is_first, bool is_last)
+{
+	const size_t ad_header_len = sizeof(data->type) + sizeof(data->data_len);
+	size_t offset = 0;
+
+	/* We may need to set the AD field data using two commands. */
+	while (offset < data->data_len) {
+		struct bt_hci_cp_le_set_ext_adv_data *set_data;
+		struct net_buf *buf;
+		int err;
+
+		buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+		if (!buf) {
+			return -ENOBUFS;
+		}
+
+		set_data = net_buf_add(buf, sizeof(*set_data));
+		(void)memset(set_data, 0, sizeof(*set_data));
+
+		set_data->handle = adv->handle;
+
+		/* Instruct the controller to minimize fragmentation. */
+		set_data->frag_pref = BT_HCI_LE_EXT_ADV_FRAG_DISABLED;
+
+		/* Determine the operation parameter value. */
+		if (is_first && (offset == 0)) {
+			set_data->op = BT_HCI_LE_EXT_ADV_OP_FIRST_FRAG;
+		} else if (is_last) {
+			/* The AD field may be split into one or two commands. */
+			if (offset != 0) {
+				/* We can always set the data in two operations
+				 * Therefore, we know that this is the last.
+				 */
+				set_data->op = BT_HCI_LE_EXT_ADV_OP_LAST_FRAG;
+			} else if (data->data_len + ad_header_len <=
+					BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN) {
+				/* First part fits. */
+				set_data->op = BT_HCI_LE_EXT_ADV_OP_LAST_FRAG;
+			} else {
+				/* The data must be split into two commands. */
+				set_data->op = BT_HCI_LE_EXT_ADV_OP_INTERM_FRAG;
+			}
+		} else {
+			set_data->op = BT_HCI_LE_EXT_ADV_OP_INTERM_FRAG;
+		}
+
+		if (offset == 0) {
+			set_data->len = MIN(data->data_len + ad_header_len,
+					    BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN);
+		} else {
+			/* No need to take min operation here,
+			 * as we can always fit the second part.
+			 */
+			set_data->len = data->data_len - offset;
+		}
+
+		if (offset == 0) {
+			set_data->data[0] = data->data_len + sizeof(data->type);
+			set_data->data[sizeof(data->data_len)] = data->type;
+			(void)memcpy(&set_data->data[ad_header_len], data->data, set_data->len);
+			offset = set_data->len - ad_header_len;
+		} else {
+			(void)memcpy(&set_data->data[0], &data->data[offset],
+			       set_data->len);
+			offset += set_data->len;
+		}
+
+		err = bt_hci_cmd_send_sync(hci_op, buf, NULL);
+		if (err) {
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+
+static int hci_set_adv_ext_fragmented(const struct bt_le_ext_adv *adv, uint16_t hci_op,
+				      const struct bt_ad *ad, size_t ad_len)
```

Comment:
Just making sure I understand this implementation:

When we send fragmented, we go through each LTV (length-type-value) structure in `ad` (the outer for loop). 
Then we iterate on each octet in `ad[i]` (the inner for loop), and call hci_set_adv_ext_fragment for each octet? 

Is this correct? 

If so, then that is an insane amount of iterations. If the host sends 10 LTV structures of 100 octets each, that's 1000 iterations! In comparison each HCI call can hold 251 octets, and can thus be done in 4 HCI commands, instead of a 1000. 

Am I incorrect?

Reply:
We are not calling set data for each octet, that would be insane. We call `hci_set_adv_ext_fragment()` for each LTV encoded AD field. Therefore, in your example we would call this command 10 times.

What can be confusing at first is that the `ad` passed into `set_ad` is a two dimentional array. The host does so internally so that it can easily add the advertising name if requested by the application.

As pointed out earlier (https://github.com/zephyrproject-rtos/zephyr/pull/37403#discussion_r692158405), this is a tradeoff between efficiency and code complexity.

Reply:
Good, then I was incorrect :D The additional layer in the `struct bt_ad` confused me. 

So we are sending one HCI command for each LTV struct - That makes sense, and would also have been my first thought to be honest. 

One issue with that approach is that the HCI command may maximum contain 251 octets, and the maximum size of an LTV struct is 256 (1 octet length field + maximum value of the length field: 255). 

So either we need to handle that (would require a more memory-heavy implementation but with fewer HCI commands), or at least add a check and throw an err/warn as well as returning an error.

Reply:
@Thalley , this PR already handles the case where the length of the LTV is greater than 251 bytes as explained in the commit message.

Reply:
@rugeGerritsen You are correct, sorry I missed that :)